src/cascadia/TerminalSettingsEditor/NullableColorPicker.cpp

```diff
@@ -0,0 +1,233 @@
+                    xaml_typename<Editor::NullableColorPicker>(),
+                    PropertyMetadata{ nullptr, PropertyChangedCallback{ &NullableColorPicker::_OnCurrentColorValueChanged } });
+        }
+        if (!_ShowNullColorButtonProperty)
+        {
+            _ShowNullColorButtonProperty =
+                DependencyProperty::Register(
+                    L"ShowNullColorButton",
+                    xaml_typename<bool>(),
+                    xaml_typename<Editor::NullableColorPicker>(),
+                    PropertyMetadata{ box_value(true) });
+        }
+        if (!_NullColorButtonLabelProperty)
+        {
+            _NullColorButtonLabelProperty =
+                DependencyProperty::Register(
+                    L"NullColorButtonLabel",
+                    xaml_typename<hstring>(),
+                    xaml_typename<Editor::NullableColorPicker>(),
+                    PropertyMetadata{ nullptr });
+        }
+        if (!_NullColorPreviewProperty)
+        {
+            _NullColorPreviewProperty =
+                DependencyProperty::Register(
+                    L"NullColorPreview",
+                    xaml_typename<Windows::UI::Color>(),
+                    xaml_typename<Editor::NullableColorPicker>(),
+                    PropertyMetadata{ box_value(Windows::UI::Colors::Transparent()) });
+        }
+    }
+
+    void NullableColorPicker::_OnCurrentColorValueChanged(const DependencyObject& d, const DependencyPropertyChangedEventArgs& /*e*/)
+    {
+        const auto& obj{ d.try_as<Editor::NullableColorPicker>() };
+        get_self<NullableColorPicker>(obj)->_UpdateColorChips();
+    }
+
+    void NullableColorPicker::_UpdateColorChips()
+    {
+        const auto& currentColor = CurrentColor();
+        for (const auto& colorChip : _colorChips)
+        {
+            const auto& chipColor = colorChip.DataContext().as<Editor::ColorTableEntry>().Color();
+            colorChip.IsChecked(currentColor ?
+                                    chipColor.R == currentColor.Value().R && chipColor.G == currentColor.Value().G && chipColor.B == currentColor.Value().B :
+                                    false);
+        }
+    }
+
+    constexpr double _CalculateLuminance(const Windows::UI::Color& color)
+    {
+        auto linearize = [](float value) {
+            value /= 255.0;
+            return (value <= 0.03928) ? (value / 12.92) : (std::pow((value + 0.055) / 1.055, 2.4));
+        };
+        double r = linearize(color.R);
+        double g = linearize(color.G);
+        double b = linearize(color.B);
+        return 0.2126 * r + 0.7152 * g + 0.0722 * b;
+    }
+
+    SolidColorBrush NullableColorPicker::CalculateBorderBrush(const Windows::UI::Color& color)
+    {
+        // Accessibility requirement is a 3:1 contrast ratio
+        static const auto whiteLuminance = _CalculateLuminance(Colors::White());
+        const auto contrastRatio = (whiteLuminance + 0.05) / (_CalculateLuminance(color) + 0.05);
```

Comment:
Instead of using relative luminance, I'd recommend copying the lambda from `Theme::IsSystemInDarkTheme()` instead (add a `static constexpr` though).

The linearization lambda above turns sRGB (roughly gamma 2.4) into gamma 1.0. Linear (1.0) gamma is needed to correctly blend two colors, because the average of two colors on a (gamma) curve doesn't make much sense. The average of two straight lines on the other hand does. But gamma'd colors more closely match what humans consider bright/dark. A 50% bright gray for our vision for instance is achieved by turning the LCD to just 20% brightness.
Due to this non-linearity, "linearizing" the sRGB color and calculating its "relative luminance" makes for a worse contrast check since it's not linear to your perception anymore. It also doesn't account for how you can easily distinguish two colors of identical luminance but different hue (e.g. blue and yellow).
It's one of the reasons why WCAG 2 isn't really that great, why it doesn't work for dark mode applications (20% value range for 50% of your vision!), and why APCA (aka WCAG 3) was created.

(FWIW, I personally remain quite skeptical of APCA and WCAG and I'd recommend avoiding them. The new formula calls it the "Lc" or "Lightness contrast", but when you look at the new formula it's just a 1.4 gamma curve. You can see that in the new formula [here](https://github.com/Myndex/SAPC-APCA/blob/master/documentation/APCA-W3-LaTeX.md) where the sRGB is incorrectly (!) converted to linear RGB (down at `Ys` where the RGB components get power'd by `Strc`) and then `f_sc(Yc)` one line above turns it back to gamma 1.4 by power'ing it with `Bclip`. That's not a problem in on itself, but if you open that document and read the "NOTICE:" at the bottom it should probably raise an eyebrow at least. If you then consider how you cannot find any information over the algorithm on the APCA website itself, despite pages upon pages of marketing speech, it should probably raise another. It almost seems like someone is trying to sell an algorithm to the W3C. I'd not use something like that tbh.)